builtin/credential/approle/path_login.go

```diff
@@ -51,14 +53,186 @@ func (b *backend) pathLoginUpdateAliasLookahead(ctx context.Context, req *logica
 // Returns the Auth object indicating the authentication and authorization information
 // if the credentials provided are validated by the backend.
 func (b *backend) pathLoginUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
-	role, roleName, metadata, _, userErr, intErr := b.validateCredentials(ctx, req, data)
-	switch {
-	case intErr != nil:
-		return nil, errwrap.Wrapf("failed to validate credentials: {{err}}", intErr)
-	case userErr != nil:
-		return logical.ErrorResponse(fmt.Sprintf("failed to validate credentials: %v", userErr)), nil
-	case role == nil:
-		return logical.ErrorResponse("failed to validate credentials; could not find role"), nil
+
+	// RoleID must be supplied during every login
+	roleID := strings.TrimSpace(data.Get("role_id").(string))
+	if roleID == "" {
+		return logical.ErrorResponse("missing role_id"), nil
+	}
+
+	// Look for the storage entry that maps the roleID to role
+	roleIDIndex, err := b.roleIDEntry(ctx, req.Storage, roleID)
+	if err != nil {
+		return nil, err
+	}
+	if roleIDIndex == nil {
+		return logical.ErrorResponse(fmt.Sprintf("invalid role_id %q", roleID)), nil
+	}
+
+	roleName := roleIDIndex.Name
+
+	roleLock := b.roleLock(roleName)
+	roleLock.RLock()
+	defer roleLock.RUnlock()
+
+	role, err := b.roleEntry(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return logical.ErrorResponse(fmt.Sprintf("invalid role_id %q", roleID)), nil
+	}
+
+	var metadata map[string]string
+	if role.BindSecretID {
+		secretID := strings.TrimSpace(data.Get("secret_id").(string))
+		if secretID == "" {
+			return logical.ErrorResponse("missing secret_id"), nil
+		}
+
+		if role.LowerCaseRoleName {
+			roleName = strings.ToLower(roleName)
+		}
+
+		secretIDHMAC, err := createHMAC(role.HMACKey, secretID)
+		if err != nil {
+			return nil, errwrap.Wrapf("failed to create HMAC of secret_id: {{err}}", err)
+		}
+
+		roleNameHMAC, err := createHMAC(role.HMACKey, roleName)
+		if err != nil {
+			return nil, errwrap.Wrapf("failed to create HMAC of role_name: {{err}}", err)
+		}
+
+		entryIndex := fmt.Sprintf("%s%s/%s", role.SecretIDPrefix, roleNameHMAC, secretIDHMAC)
+
+		secretIDLock := b.secretIDLock(secretIDHMAC)
+		secretIDLock.RLock()
+		secretIDUnlockFunc := secretIDLock.RUnlock
+		defer func() {
```

Comment:
Honestly, I'd prefer if you didn't do this.  In the two return scenarios below you could just unlock manually. Then when you're in the switch, in the first case you know it'll stay a read lock and you can immediately defer the RUnlock; in the second case you know you'll need to switch it so you can runlock, lock, immediately defer the write lock.